### ISSUE TITLE
Add a CRUD example to Getting Started doc

### ISSUE THREAD

**Issue by reinert** [efcore/EFCore.FSharp#54]
First of all, thank you very much for this project.

Very useful.

I'm a newcomer in the F# world and I'm trying to use EF along with F# models. I could make trough every step in the Getting Started section successfully. But I do not have enough knowledge to put pieces together yet. So I would like to ask you to add a simple CRUD example with the created Post model in the Getting Started article.

Thank you.

**Comment by simontaite** [comment]
Hi @reinert,

Thank you for the feedback, I'm glad you're finding the project useful
A CRUD example is something that we definitely need. I'm happy to look at creating one, I've been meaning to add and expand documentation for the project as I can

I can't offer an exact timeframe on it at the moment but it's definitely something I'd want to add sooner rather than later

**Comment by reinert** [comment]
@simontaite Thank you very much for your reply!

Could you at least put a few lines of code here showing how to instantiate a Context and inserting a Blog Post? As also getting all records from database.

I really only need it for now.

Please :)

**Comment by reinert** [comment]
I tried the following, but no success:

    using (new BloggingContext()) (fun ctx ->
        let b = ctx.Blogs.Find(1)
        printfn "%s" b.Url
    ) |> ignore

**Comment by simontaite** [comment]
To instantiate a context and then query it you can use something like
```fsharp
use context = new BlogContext()
let blogPosts = context.BlogPosts|> Seq.toList
```

Don't forget though that even though it's not very functional, if were an MVC controller for example, you would also be able to inject the context as a constructor parameter and let the framework handle creating the context for you

To insert an entry you would want to do something like
```fsharp
let blogPost = { Id = 0; Slug = "my-slug"; (*etc etc*) }

// Add to context
blogPost |> context.BlogPosts.Add |> ignore
context.SaveChanges() |> ignore
```

It's a little trickier to update an entity because of the immutable nature of F# records
Working on ways to ease this but at the moment, I'm using this sort of pattern

```fsharp
// Separate function
let updateBlogPost (context: BlogContext) (entity: BlogPost) = 
        let currentEntry = context.BlogPosts.Find(entity.Id)
        context.Entry(currentEntry).CurrentValues.SetValues(entity)
        context.SaveChanges |> ignore

// Where we have our blog post we want to update
let blogPost = { Id = 0; Slug = "my-slug"; (*etc etc*) }
let updatedBlogPost = { blogPost with Slug = "my-updated-slug" }

updateBlogPost context updatedBlogPost
```

Part of the plan to make this project more usable is to abstract over a lot of the boilerplate needed for updating records like that but that's definitely further down the list

The examples above should work, it's been a little while since I've had time to work on this, I'm currently rebuilding it locally to double check

**Comment by reinert** [comment]
Thanks @simontaite !

I notice that context.Posts DbSet is null after instantiating the Context.

I guess this is related to the IoC.

Do you know how could I get an instance of the DbContext from Service provider?

**Comment by reinert** [comment]
There is a problem in declaring DbSet as members.

This [SO answer](https://stackoverflow.com/a/43357781) helped me solve it.

**Comment by simontaite** [comment]
Thanks, I'll make sure to update the documentation to include that

**Comment by reinert** [comment]
This is my BloggingContext working version:

```fsharp
type BloggingContext() =  
    inherit DbContext()

    [<DefaultValue>]
    val mutable m_blogs : DbSet<Blog>
    member __.Blogs with get() = __.m_blogs 
                                and set v = __.m_blogs <- v
    [<DefaultValue>]
    val mutable m_posts : DbSet<Post>
    member __.Posts with get() = __.m_posts 
                                and set v = __.m_posts <- v

    override __.OnConfiguring(options: DbContextOptionsBuilder) : unit =
        options.UseSqlite("Data Source=blogging.db") |> ignore
```

**Comment by reinert** [comment]
A simple working CRUD:

```fsharp
open System
open BloggingModel
open Microsoft.EntityFrameworkCore

let getBlogs (ctx : BloggingContext) =
    ctx.Blogs |> Seq.toList

let createBlog (ctx : BloggingContext) (b : Blog) =
    ctx.Blogs.Add b |> ignore
    ctx.SaveChanges() |> ignore
    b

let updateBlog (ctx: BloggingContext) (b: Blog) = 
    let currentEntry = ctx.Blogs.Find b.Id
    ctx.Entry(currentEntry).CurrentValues.SetValues(b)
    ctx.SaveChanges() |> ignore
    b

let deleteBlog (ctx: BloggingContext) (b: Blog) =
    ctx.Blogs.Remove b |> ignore
    ctx.SaveChanges() |> ignore
    b

let clearTable (ctx : BloggingContext) (table : string) =
    ctx.Database.ExecuteSqlRaw("DELETE from " + table)

[<EntryPoint>]
let main argv =
    use ctx = new BloggingContext()

    clearTable ctx "Blogs" |> ignore

    // create an entity with specified Id
    let medium = createBlog ctx { Id = 11; Url = "medium.com" }

    // create an entity with auto-increment Id
    let so = createBlog ctx { Id = 0; Url = "stackoverflow.com" }

    // print all records
    getBlogs ctx |> printfn "%A"

    // update stackoverflow to stackexchange
    let se = updateBlog ctx { Id = so.Id; Url = "stackexchange.com" }
    
    getBlogs ctx |> printfn "%A"

    // delete first blog
    deleteBlog ctx medium |> ignore

    getBlogs ctx |> printfn "%A"

    ctx.Dispose() // not sure if this is necessary

    0

```

**Comment by simontaite** [comment]
Hi @reinert 
That's a great example! Thank you so much.
You ok with me using this in our documentation?

**Comment by reinert** [comment]
Yes, sure! I really appreciate your awesome work in this project! It's a pleasure to contribute!

**Comment by GunnerGuyven** [comment]
The `updateBlogPost` function can be made generic with some prep work:
1. create an interface for retrieving the PK value
```fsharp
type ITable =
    abstract ID : int
```
2. have your table type use the interface
```fsharp
[<CLIMutable>]
type BlogPost =
    { [<Key>]
      ID: int
      [<MaxLength(30)>]
      Slug: string }
    interface ITable with
        member x.ID = x.ID
```
3. Redefine update in this way
```fsharp
let updateEntity (ctx: DbContext) (entity: 'E :> ITable) =
    let current = ctx.Find<'E> entity.ID

    ctx
        .Entry(current)
        .CurrentValues.SetValues(entity :> obj)
```
The benefit to this is that you don't need an object-specific update methods scattered around.  This is working for me so far and I don't see a problem with it yet.  Putting it here in case it helps others who would like to have just the one `update` in their codebase.

**Comment by simon-reynolds** [comment]
Thanks @GunnerGuyven 

I like your `ITable` solution though, will definitely add it to the documentation.
Even though F# does also allow member constraints on properties, an interface is more explicit

The project also has a generic version of `updateEntity` [here](https://github.com/efcore/EFCore.FSharp/blob/master/src/EFCore.FSharp/DbContextHelpers.fs)

`updateEntity (ctx: #DbContext) (key: 'a -> obj) (entity : 'a when 'a : not struct)`
This is more generic, requiring a function to identify the key, so for example, you would use
`updateEntity ctx (fun b -> b.Id :> obj) b`

This is intentionally kept as generic as possible to avoid forcing a constraint or pattern on users.
Only caveat is that the implementation will be changing soon to remove the call to `SaveChanges`, I'm not happy having that there when a user may not be expecting to save at that point.

**Comment by GunnerGuyven** [comment]
I'm glad to hear about this, because I didn't find that method in my searching.  The `SaveChanges` is indeed a deal-breaker for me because i'm doing a macro operation and I want to apply a number of entity updates before saving them all together.  I'm both glad you're intending to change that, and also happy to see that signature: it has the benefit of allowing composite keys (I have one in my context and was worrying how I'd adapt it to the common interface).  I was also considering passing through the entity for having potentially multi-stage changes over time without having to track externally, seeing it here gives me enough validation that its a decent idea that I'll adopt that signature too :+1: 

Also, thank you for telling me that F# _does_ support member constraints.  I had been searching for that when I happened on type interfaces and thought "OK, I can make that work, then".  I'll do some more research, and also adopt the delegate function for identifying key because I like it :)

Thank you very much!

**Comment by simon-reynolds** [comment]
Hi @GunnerGuyven 
I've published a new version (5.0.3-alpha2) to NuGet, includes a whole lot of DbContext methods in the `EntityFrameworkCore.FSharp.DbContextHelpers` module

* `updateEntity` no longer calls `SaveChanges`
* has some other more F#-ish wrappers around `DbContext` including a `tryFindEntity` that returns an `option`
* Async versions of all that use `async { ... }`

There's some example usage at https://efcore.github.io/EFCore.FSharp/How_Tos/Use_DbContextHelpers.html

**Comment by GunnerGuyven** [comment]
This is excellent!  I have one small complaint.  The signature:
```fsharp
let updateEntity (ctx: DbContext) (key: 'a -> obj) (entity : 'a when 'a : not struct) =
```
requires the usage:
```fsharp
updateEntity ctx (fun b -> b.Id :> obj) modifiedBlogPost
```
but if you altered it to:
```fsharp
let updateEntity (ctx: DbContext) (key: 'a -> 'b) (entity : 'a when 'a : not struct) =
```
you'd save the consumer the cast to `obj`:
```fsharp
updateEntity ctx (fun b -> b.Id) modifiedBlogPost
```
I realize this may not be feasible due to things happening elsewhere in your API, I haven't tried to make it work except in my own code so far (and it seems to).
